### DISCUSSION TITLE
Exploring taipy's Suitability for State Management in a RPC/Web UI Service

### DISCUSSION THREAD

**Discussion by mosmuell** [Avaiga/taipy#227] [Q&A] [answered]
Hello,

Firstly, I would like to express my gratitude for your excellent work on the taipy project. I have been experimenting with your package, and I'm considering whether it aligns with my specific use case. I thought it would be best to ask here to ensure I'm not overlooking any potential solutions.

To provide some context, I am developing a package for services that exposes an RPC server and a web UI. In this package, the user is expected to implement a class, the public attributes and methods of which are then exposed. Clients can connect to the service to invoke these methods or modify the class attributes. Essentially, the class encapsulates the business logic of my service.
I have a set of requirements that I need to fulfil:
1. The service should be capable of executing certain tasks periodically (e.g. reading out a sensor and logging it to a database, implemented with asyncio).
2. These tasks should have access to the current values of the class attributes (e.g. for changing the frequency of the readout)
3. There should be a single "state", ensuring that all clients connecting to the service see the same attribute values.
4. The web interface should always display the most recent data (e.g. it should always reflect the actual value of the sensor readout)

Given these requirements, I am trying to ascertain whether taipy GUI is a suitable solution for my project. I am particularly interested in understanding how taipy handles state management. From my understanding, each connection to the web UI maintains its own set of variables along with their values. This seems to suggest that I cannot have a global state which is shared among all connections. I am wondering if it would be possible to have the web UI bind to the attributes of the class, which can then be updated through both web UI interactions and RPC calls?

I would greatly appreciate any insights or advice you could provide on this matter.

Thank you in advance for your time and assistance.

**Comment by FlorianJacta** [comment, marked as answer]
Hi,

Thank you for your kind words and for considering Taipy for your project. We would happily provide insights into how Taipy handles state management and whether it aligns with your requirements.

Taipy offers a way to execute a function periodically. Using [long callbacks](https://docs.taipy.io/en/latest/manuals/gui/callbacks/#task-progress), you are able to call a function with a frequency and execute functions depending on what time it is. We could provide more code to satisfy your needs.

Taipy also has a built-in data and pipeline management system that allows you to manage your tasks and data quickly and efficiently. It is referred to as Taipy Core in the documentation, and we might add a CRON directly to it in the future.

Regarding "state" management, by default, Taipy follows a request-response model where each client connection maintains its own set of variables and values. You could use the "single_client" mode where only one state exists, and each user has the same state. However, users will have the same view and interaction with the application.

Three solutions:

- Use the same [long callbacks](https://docs.taipy.io/en/latest/manuals/gui/callbacks/#task-progress) mentioned before to update each state with the global value of the application. This will synchronize the application for every user.

- Use the backend of Taipy. You will be able to manage and store your data. This will let anyone have the same data as the basis of their application. You can use on top of that the [long callbacks](https://docs.taipy.io/en/latest/manuals/gui/callbacks/#task-progress).

- Update the class attributes based on user interactions in the web UI and through RPC calls by establishing appropriate event handling.

Moreover, the concept of shared variables will be present in the next release for 3.0. It will be released in early october.

Thank you for considering Taipy, and we wish you the best in your project. If you have any further questions, please don't hesitate to ask.
